crates/neon/src/macros.rs

```diff
@@ -0,0 +1,148 @@
+//! Helper module to add documentation to macros prior to re-exporting.
+
+/// Marks a function as the main entry point for initialization in
+/// a Neon module.
+///
+/// This attribute should only be used _once_ in a module and will
+/// be called each time the module is initialized in a context.
+///
+/// If a `main` function is not provided, all registered exports will be exported.
+///
+/// ```
+/// # use neon::prelude::*;
+/// # fn main() {
+/// #[neon::main]
+/// fn main(mut cx: ModuleContext) -> NeonResult<()> {
+///     // Export all registered exports
+///     neon::registered().export(&mut cx)?;
+///
+///     let version = cx.string("1.0.0");
+///
+///     cx.export_value("version", version)?;
+///
+///     Ok(())
+/// }
+/// # }
+/// ```
+pub use neon_macros::main;
+
+/// Register an item to be exported by the Neon addon
+///
+/// ## Exporting constants and statics
+///
+/// ```
+/// #[neon::export]
+/// static GREETING: &str = "Hello, Neon!";
+///
+/// #[neon::export]
+/// const ANSWER: u8 = 42;
+/// ```
+///
+/// ### Renaming an export
+///
+/// By default, items will be exported with their Rust name. Exports may
+/// be renamed by providing the `name` attribute.
+///
+/// ```
+/// #[neon::export(name = "myGreeting")]
+/// static GREETING: &str = "Hello, Neon!";
+/// ```
+///
+/// ### JSON exports
+///
+/// Complex values may be exported by automatically serializing to JSON and
+/// parsing in JavaScript. Any type that implements `serde::Serialize` may be used.
+///
+/// ```
+/// #[neon::export(json)]
+/// static MESSAGES: &[&str] = &["hello", "goodbye"];
+/// ```
+///
+/// ## Exporting functions
+///
+/// Functions may take any type that implements [`TryFromJs`](crate::types::extract::TryFromJs) as
+/// an argument and return any type that implements [`TryIntoJs`](crate::types::extract::TryIntoJs).
+///
+/// ```
+/// #[neon::export]
+/// fn add(a: f64, b: f64) -> f64 {
+///     a + b
+/// }
+/// ```
+///
+/// ### Interact with the JavaScript runtime
```

Comment:
I chose to add an "Advanced" section to the bottom instead since most users won't need to worry about this. It also gave me a good place to co-locate docs with the `result` attribute I added.